lib/smack/SmackInstGenerator.cpp

```diff
@@ -910,6 +910,84 @@ void SmackInstGenerator::visitIntrinsicInst(llvm::IntrinsicInst& ii) {
     else
       generateUnModeledCall(ci);
   };
+  
+  // Count leading zeros
+  static const auto ctlz = [this](CallInst* ci) {
+    if (SmackOptions::BitPrecise) {
+      auto width = ci->getArgOperand(0)->getType()->getIntegerBitWidth();
+      auto var = rep->expr(ci->getArgOperand(0));
+
+      // e.g., if v[32:31] == 1 then 0bv32 else if v[31:30] == 1 then 1bv32 else
+      // ... else if v[1:0] == 1 then 31bv32 else 32bv32
+      const Expr* body = Expr::lit(width, width);
+      for (unsigned i = 0; i < width; ++i) {
+        body = Expr::if_then_else(Expr::eq(Expr::bvExtract(var, i+1, i),
+                                           Expr::lit(1,1)),
+                                  Expr::lit(width-i-1, width),
+                                  body);
+      }
+
+      // Handle the is_zero_undef case, i.e. if the flag is set and the argument
+      // is zero, then the result is undefined.
+      auto isZeroUndef = rep->expr(ci->getArgOperand(1));
+      body = Expr::if_then_else(Expr::and_(Expr::eq(isZeroUndef, Expr::lit(1, 1)),
+                                           Expr::eq(var, Expr::lit(0,width))),
+                                rep->expr(ci), // The result is undefined
+                                body);
+      emit(Stmt::assign(rep->expr(ci), body));
+    }
+    else
+      generateUnModeledCall(ci);
+  };
+
+  // Count trailing zeros
+  static const auto cttz = [this](CallInst* ci) {
+    if (SmackOptions::BitPrecise) {
+      auto width = ci->getArgOperand(0)->getType()->getIntegerBitWidth();
+      auto arg = rep->expr(ci->getArgOperand(0));
+
+      // e.g., if v[1:0] == 1 then 0bv32 else if v[2:1] == 1 then 1bv32 else
+      // ... else if v[32:31] == 1 then 31bv32 else 32bv32
+      const Expr* body = Expr::lit(width, width);
+      for (unsigned i = width; i > 0; --i) {
+        body = Expr::if_then_else(Expr::eq(Expr::bvExtract(arg, i, i-1),
+                                           Expr::lit(1,1)),
+                                  Expr::lit(i-1, width),
+                                  body);
+      }
+
+      // Handle the is_zero_undef case, i.e. if the flag is set and the argument
+      // is zero, then the result is undefined.
+      auto isZeroUndef = rep->expr(ci->getArgOperand(1));
+      body = Expr::if_then_else(Expr::and_(Expr::eq(isZeroUndef, Expr::lit(1, 1)),
+                                           Expr::eq(arg, Expr::lit(0,width))),
+                                rep->expr(ci), // The result is undefined
```

Comment:
I couldn't find a way to create a nondet value directly using our functionality. This assigns the variable to itself, which is fine so long as it hasn't been assigned previously. I'd like something more predictable.

Reply:
Inserted a havoc to make sure the value is unknown.